File: server/common/protocol.py
import socket
import struct
from types import SimpleNamespace
from typing import Optional, Tuple
from common.utils import Bet

# Constants for message types and protocol errors
consts = SimpleNamespace(
    MSG_SUCCESS=0x00,  # Success message
    MSG_BET=0x01,  # Bet message
    MSG_ECHO=0x02,  # Echo message
    MSG_ERROR=0x03,  # Error message
    REJECT_MALFORMED=0x04,  # Malformed message rejection
    REJECT_INVALID=0x05,  # Invalid message rejection
    MAX_MSG_LENGTH=1024,  # Maximum message length
    MIN_PROTOCOL_MESSAGE_LENGTH=6  # Minimum protocol message length
)


class ProtocolError(Exception):
    def __init__(self, message: str, code: int) -> None:
        super().__init__(message)
        self.code: int = code


class Protocol:
    def __init__(self, conn: socket.socket):
        self.conn: socket.socket = conn
        self.consts: SimpleNamespace = consts  # Include the consts inside the class

    def read_exactly(self, n: int) -> bytes:
        """Ensures that exactly n bytes are read from the socket."""
        data: bytearray = bytearray()
        while len(data) < n:
            packet: Optional[bytes] = self.conn.recv(n - len(data))
            if not packet:
                raise ConnectionError("Short read")
            data.extend(packet)
        return bytes(data)

    def read_with_timeout(self, n: int, timeout: float = 0.1) -> bytes:
        """Reads up to n bytes with a timeout to avoid blocking indefinitely."""
        data: bytearray = bytearray()
        self.conn.settimeout(timeout)
        try:
            while len(data) < n:
                packet: Optional[bytes] = self.conn.recv(n - len(data))
                if not packet:
                    break
                data.extend(packet)
        except socket.timeout:
            pass  # Stop reading if timeout occurs
        finally:
            self.conn.settimeout(None)  # Reset the timeout to default (blocking mode)
        return bytes(data)

    def send_all(self, data: bytes) -> None:
        """Ensures that all data is sent over the socket."""
        total_sent: int = 0
        while total_sent < len(data):
            sent: int = self.conn.send(data[total_sent:])
            if sent == 0:
                raise ConnectionError("Socket connection broken")
            total_sent += sent

    def receive_message(self) -> bytes:
        """Reads the length header and the entire message."""
        header: bytes = self.conn.recv(4, socket.MSG_PEEK)  # Peek to check if it's a valid header

        if all(32 <= byte <= 126 for byte in header):
            raise ProtocolError("Error unpacking header", self.consts.MSG_ECHO)

        # Otherwise, it's likely a structured protocol message
        header = self.read_exactly(4)
        message_length: int = struct.unpack('>I', header)[0]

        if message_length < self.consts.MIN_PROTOCOL_MESSAGE_LENGTH or message_length > self.consts.MAX_MSG_LENGTH:
            raise ProtocolError("Invalid message length", self.consts.REJECT_MALFORMED)

        message_data: bytes = self.read_exactly(message_length)
        return message_data

    def send_response(self, code: int, message: str) -> None:
        """Sends a response to the client with a status code and message."""
        message_length: int = 4 + len(message) + 1  # HEADER_LENGTH + message + 1 byte for code
        header: bytes = struct.pack('>I', message_length)
        response: bytes = header + bytes([code]) + message.encode('utf-8')
        self.send_all(response)

    def handle_echo_message(self, initial_data: bytes) -> None:
        """Handles echo messages by reading the rest of the message and sending it back."""
        remaining_length = self.consts.MAX_MSG_LENGTH - len(initial_data)
        remaining_data = self.read_with_timeout(remaining_length)
        self.send_all(initial_data + remaining_data)

    def deserialize_bet(self, data: bytes) -> Bet:
        """Deserializes a Bet object from binary data according to the protocol."""
        offset: int = 0

        # Deserialize Agency (4 bytes, uint32)
        agency: int = struct.unpack('>I', data[offset:offset + 4])[0]
        offset += 4

        # Deserialize First Name (variable length string)
        first_name, offset = self.deserialize_string(data, offset)

        # Deserialize Last Name (variable length string)
        last_name, offset = self.deserialize_string(data, offset)

        # Deserialize Document (variable length string)
        document, offset = self.deserialize_string(data, offset)

        # Deserialize Birth Date (10 bytes string)
        birthdate: str = data[offset:offset + 10].decode('utf-8')
        offset += 10

        # Deserialize Number (4 bytes, uint32)
        number: int = struct.unpack('>I', data[offset:offset + 4])[0]

        return Bet(str(agency), first_name, last_name, document, birthdate, str(number))

    def deserialize_string(self, data: bytes, offset: int) -> Tuple[str, int]:
        """Deserializes a string from binary data, reading the length prefix first."""
        str_length: int = struct.unpack('>I', data[offset:offset + 4])[0]
        offset += 4
        string: str = data[offset:offset + str_length].decode('utf-8')
        offset += str_length
        return string, offset

File: client/common/protocol.go
package common

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"net"
)

// MsgType is an enumeration of the different message types and protocol errors
type MsgType int

const (
	MSG_SUCCESS      MsgType = 0x00 // 0x00, Success message
	MSG_BET          MsgType = 0x01 // 0x01, Bet message
	MSG_ECHO         MsgType = 0x02 // 0x02, Echo message
	MSG_ERROR        MsgType = 0x03 // 0x03, Error message
	REJECT_MALFORMED MsgType = 0x04 // 0x04, Malformed message rejection
	REJECT_INVALID   MsgType = 0x05 // 0x05, Invalid message rejection
)

func (m MsgType) String() string {
	switch m {
	case REJECT_MALFORMED:
		return "REJECT_MALFORMED"
	case REJECT_INVALID:
		return "REJECT_INVALID"
	case MSG_SUCCESS:
		return "MSG_SUCCESS"
	case MSG_BET:
		return "MSG_BET"
	case MSG_ECHO:
		return "MSG_ECHO"
	case MSG_ERROR:
		return "MSG_ERROR"
	default:
		return "UNKNOWN"
	}
}

// Protocol defines the behavior of our protocol
type Protocol struct {
	conn net.Conn
}

// NewProtocol creates a new instance of the protocol
func NewProtocol(conn net.Conn) *Protocol {
	return &Protocol{conn: conn}
}

// SerializeBet serializes a Bet object into binary format according to the protocol.
func (p *Protocol) SerializeBet(bet Bet) ([]byte, error) {
	var buffer bytes.Buffer

	// Serialize Agency (4 bytes, uint32)
	if err := binary.Write(&buffer, binary.BigEndian, uint32(bet.Agency)); err != nil {
		return nil, err
	}

	// Serialize FirstName (4 bytes length prefix + string)
	if err := p.serializeString(&buffer, bet.FirstName); err != nil {
		return nil, err
	}

	// Serialize LastName (4 bytes length prefix + string)
	if err := p.serializeString(&buffer, bet.LastName); err != nil {
		return nil, err
	}

	// Serialize Document (4 bytes length prefix + string)
	if err := p.serializeString(&buffer, bet.Document); err != nil {
		return nil, err
	}

	// Serialize BirthDate (10 bytes, string "YYYY-MM-DD")
	birthDateStr := bet.BirthDate.Format("2006-01-02")
	if len(birthDateStr) != 10 {
		return nil, fmt.Errorf("invalid birth date format")
	}
	if _, err := buffer.WriteString(birthDateStr); err != nil {
		return nil, err
	}

	// Serialize Number (4 bytes, uint32)
	if err := binary.Write(&buffer, binary.BigEndian, uint32(bet.Number)); err != nil {
		return nil, err
	}

	return buffer.Bytes(), nil
}

// serializeString serializes a string with a 4-byte length prefix.
func (p *Protocol) serializeString(buffer *bytes.Buffer, str string) error {
	strLength := uint32(len(str))
	if err := binary.Write(buffer, binary.BigEndian, strLength); err != nil {
		return err
	}
	if _, err := buffer.Write([]byte(str)); err != nil {
		return err
	}
	return nil
}

// SendAll ensures that all data is sent over the socket
func (p *Protocol) SendAll(data []byte) error {
	totalSent := 0
	for totalSent < len(data) {
		sent, err := p.conn.Write(data[totalSent:])
		if err != nil {
			return err
		}
		if sent == 0 {
			return fmt.Errorf("socket connection broken")
		}
		totalSent += sent
	}
	return nil
}

// ReadExactly ensures that exactly n bytes are read from the socket
func (p *Protocol) ReadExactly(n int) ([]byte, error) {
	data := make([]byte, n)
	totalRead := 0
	for totalRead < n {
		read, err := p.conn.Read(data[totalRead:])
		if err != nil {
			return nil, err
		}
		if read == 0 {
			return nil, fmt.Errorf("socket connection broken")
		}
		totalRead += read
	}
	return data, nil
}

// SendBet sends a serialized Bet object to the server.
func (p *Protocol) SendBet(bet Bet) error {
	betData, err := p.SerializeBet(bet)
	if err != nil {
		log.Errorf("Failed to serialize bet: %v", err)
		return err
	}

	messageLength := uint32(len(betData)) + 1 // +1 for the message type
	header := make([]byte, 4)
	binary.BigEndian.PutUint32(header, messageLength)

	// Send the message length, type, and serialized data
	if err := p.SendAll(header); err != nil {
		log.Errorf("Failed to send message length: %v", err)
		return err
	}

	messageType := byte(MSG_BET) // Use constant for MSG_BET
	if err := p.SendAll([]byte{messageType}); err != nil {
		log.Errorf("Failed to send message type: %v", err)
		return err
	}

	if err := p.SendAll(betData); err != nil {
		log.Errorf("Failed to send bet data: %v", err)
		return err
	}

	return nil
}

// ReceiveResponse receives and parses the server's response
func (p *Protocol) ReceiveResponse() (int, string, error) {
	header, err := p.ReadExactly(4)
	if err != nil {
		return 0, "", fmt.Errorf("failed to read response header: %v", err)
	}

	messageLength := binary.BigEndian.Uint32(header)
	statusCode, err := p.ReadExactly(1)
	if err != nil {
		return 0, "", fmt.Errorf("failed to read status code: %v", err)
	}

	messageBody, err := p.ReadExactly(int(messageLength) - 5)
	if err != nil {
		return 0, "", fmt.Errorf("failed to read message body: %v", err)
	}

	return int(statusCode[0]), string(messageBody), nil
}
